aten/src/ATen/native/quantized/cpu/qlinear.cpp

```diff
@@ -900,6 +901,104 @@ at::Tensor PackedLinearWeightsOnednn:: apply_tanh(
+    torch::List<double>& post_op_args,
+    std::string& post_op_algorithm) {
+  using ideep::tensor;
+  const int64_t dim = input.dim();
+  TORCH_CHECK(input.scalar_type() == c10::ScalarType::Byte,
+      "qlinear with mkldnn tensor: data type of input should be uint8 (unsigned char).");
+  TORCH_CHECK(onednn_weight.scalar_type() == c10::ScalarType::Char,
+      "qlinear with mkldnn tensor: data type of weight should be int8 (char).");
+  TORCH_CHECK(
+      weight_scales.scalar_type() == c10::ScalarType::Float, "weight scales should be dtype c10::ScalarType::Float.");
+  if (fp32_output) {
+    TORCH_CHECK(
+        output_scale == 1.0f && output_zero_point == 0, "onednn qlinear: expect scale=1 and zero point=0 for fp32 output");
+  }
+
+  auto src = at::native::itensor_from_tensor(input);
+  auto packed_weight = at::native::itensor_from_mkldnn(onednn_weight);
+  int64_t K = input.size(dim - 1), M = input.numel() / K, N = packed_weight.get_dim(1);
+  c10::optional<ideep::tensor> onednn_bias{c10::nullopt};
+  bool with_bias = bias.has_value();
+  if (with_bias) {
+    if (bias.value().dim() == 1) {
+      auto b_reshape = bias.value().reshape({1, bias.value().size(0)});
+      onednn_bias = at::native::itensor_view_from_dense(b_reshape);
+    } else {
+      onednn_bias = at::native::itensor_view_from_dense(bias.value());
+    }
+  }
+  std::vector<int64_t> src_dims = {M, K};
+  std::vector<int64_t> dst_dims = {M, N};
+  auto output_dtype = fp32_output ? c10::kFloat : c10::kByte;
+  at::Tensor output = at::empty(
+    dst_dims,
+    device(c10::kCPU)
+        .dtype(output_dtype)
+  );
+  if (output.numel() == 0) {
+    return output;
+  }
+  tensor dst = at::native::itensor_view_from_dense(output);
+
+  // Create onednn primitive
+  auto src_desc = tensor::desc(src_dims, ideep::data_type::u8, ideep::format_tag::any);
+  auto weights_desc = packed_weight.get_desc();
+  auto dst_dtype = fp32_output ? ideep::data_type::f32 : ideep::data_type::u8;
+  auto dst_desc = tensor::desc(dst_dims, dst_dtype, ideep::format_tag::any);
+  auto bias_desc = with_bias ?
+      tensor::desc(onednn_bias.value().get_dims(), ideep::data_type::f32, ideep::format_tag::any) :
+      tensor::desc();
+  auto op_attr = onednn_utils::create_attr_by_post_op(post_op_name, post_op_args);
+  op_attr.set_scales_mask(DNNL_ARG_SRC, 0);
+  op_attr.set_zero_points_mask(DNNL_ARG_SRC, 0);
+  op_attr.set_scales_mask(DNNL_ARG_WEIGHTS, ideep::utils::op_scale_mask(weight_scales.numel()));
+  op_attr.set_scales_mask(DNNL_ARG_DST, 0);
+  op_attr.set_zero_points_mask(DNNL_ARG_DST, 0);
+  op_attr.set_scratchpad_mode(dnnl::scratchpad_mode::user);
+  auto engine = ideep::engine::cpu_engine();
+  auto primitive_desc = with_bias ?
+      dnnl::matmul::primitive_desc(engine, src_desc, weights_desc, bias_desc, dst_desc, op_attr) :
+      dnnl::matmul::primitive_desc(engine, src_desc, weights_desc, dst_desc, op_attr);
+  auto primitive = dnnl::matmul(primitive_desc);
```

Comment:
Do we need to check if we need to do weight reorder here?

Reply:
Thanks. Check is added.